### PR TITLE
[Time Conversion] Timestamp to time

### PR DESCRIPTION
**Pull request by Devjiu**
Adding timestamp to time conversion. The  units is also taken into account.

Resolves: https://github.com/intel-ai/hdk/issues/318

Signed-off-by: Dmitrii Makarenko <dmitrii.makarenko@intel.com>

### CODE REVIEW

File: omniscidb/Tests/QueryBuilderTest.cpp [hunk 1]
```diff
@@ -2007,17 +2111,19 @@ TEST_F(QueryBuilderTest, CastTimestampExpr) {
   checkCast(scan.ref("col_timestamp").cast("int16"), ctx().int16());
   checkCast(scan.ref("col_timestamp").cast("int32"), ctx().int32());
   checkCast(scan.ref("col_timestamp").cast("int64"), ctx().int64());
-  checkCast(scan.ref("col_timestamp").cast("fp32"), ctx().fp32());
-  checkCast(scan.ref("col_timestamp").cast("fp64"), ctx().fp64());
-  checkCast(scan.ref("col_timestamp").cast("dec(10,2)"), ctx().decimal(8, 10, 2));
```
Comment (Devjiu) [inline omniscidb/Tests/QueryBuilderTest.cpp:2012]: Should we be able to cast to decimal? This cast looks a bit weird to support.

File: omniscidb/Tests/QueryBuilderTest.cpp [hunk 2]
```diff
@@ -362,17 +366,45 @@ class QueryBuilderTest : public TestSuite {
 
     createTable("withNull", {{"a", ctx().int64()}});
     insertCsvValues("withNull", "1\nNULL");
+
+    createTable("test_tmstmp",
+                {{"col_bi", ctx().int64()},
+                 {"col_tmstp", ctx().timestamp(hdk::ir::TimeUnit::kSecond)},
```
Comment (ienkovich) [inline omniscidb/Tests/QueryBuilderTest.cpp:372]: Also, try non-nullable types. For nullable types add some NULL values.

File: omniscidb/Tests/QueryBuilderTest.cpp (hunk 1 above)
Comment (Devjiu) [reply omniscidb/Tests/QueryBuilderTest.cpp:2013]: Yes, we are removing such cast.

**Comment by Devjiu** [comment]
I am not sure, that nulls handled in correct way. Arrow array in case Time32[s] will have int32, canonical type is int64, so in ResultSet will be null int64_t, after conversion to arrow int32 null.

File: omniscidb/QueryEngine/CastIR.cpp [hunk 3]
```diff
@@ -108,6 +108,14 @@ llvm::Value* CodeGenerator::codegenCast(llvm::Value* operand_lv,
           operand_type->as<hdk::ir::TimestampType>()->unit(),
           type->nullable());
     }
+    if (operand_type->isTimestamp() && type->isTime()) {
+      const auto operand_unit = (operand_type->isTimestamp())
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: removed

File: omniscidb/QueryEngine/CastIR.cpp [hunk 4]
```diff
@@ -108,6 +108,14 @@ llvm::Value* CodeGenerator::codegenCast(llvm::Value* operand_lv,
...
+                                    ? operand_type->as<hdk::ir::TimestampType>()->unit()
+                                    : hdk::ir::TimeUnit::kSecond;
+      // if (operand_unit != type->as<hdk::ir::TimestampType>()->unit()) {
+      return codegenCastTimestampToTime(operand_lv, operand_type, type, type->nullable());
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: removed unnecessary argument.

File: omniscidb/QueryEngine/CastIR.cpp [hunk 5]
```diff
@@ -159,6 +167,60 @@ llvm::Value* CodeGenerator::codegenCastTimestampToDate(llvm::Value* ts_lv,
...
+  AUTOMATIC_IR_METADATA(cgen_state_);
+  auto& ctx = hdk::ir::Context::defaultCtx();
+  const auto operand_unit = operand_type->as<hdk::ir::TimestampType>()->unit();
+  const auto target_unit = target_type->as<hdk::ir::TimeType>()->unit();
+  LOG(ERROR) << "casting from: " << operand_type->toString()
+             << " to: " << target_type->toString();
+  LOG(ERROR) << "  units from: " << operand_unit << " to: " << target_unit;
+  CHECK(ts_lv->getType()->isIntegerTy(64));
+  const auto target_sec_scale = hdk::ir::unitsPerSecond(target_unit) /
+                                hdk::ir::unitsPerSecond(hdk::ir::TimeUnit::kSecond);
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: removed

File: omniscidb/QueryEngine/CastIR.cpp [hunk 6]
```diff
@@ -159,6 +167,60 @@ llvm::Value* CodeGenerator::codegenCastTimestampToDate(llvm::Value* ts_lv,
...
+  LOG(ERROR) << " op unit >= target unit ( " << hdk::ir::unitsPerSecond(operand_unit)
+             << "/" << hdk::ir::unitsPerSecond(target_unit) << " ) scale: " << scale
+             << " taget sec scale: " << target_sec_scale;
+  // codegenCastBetweenIntTypesOverflowChecks(ts_lv, operand_type, target_type, scale);
+  if (!nullable) {
+    return cgen_state_->emitCall("TimestampTruncateHighPrecisionToSmallerTime",
+                                 {{ts_lv,
+                                   cgen_state_->llInt(scale),
+                                   cgen_state_->llInt(target_sec_scale),
+                                   cgen_state_->inlineIntNull(ctx.int64())}});
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: removed

File: omniscidb/QueryEngine/CastIR.cpp [hunk 7]
```diff
@@ -159,6 +167,60 @@ llvm::Value* CodeGenerator::codegenCastTimestampToDate(llvm::Value* ts_lv,
...
+  }
+  return cgen_state_->emitCall("TimestampTruncateHighPrecisionToSmallerTimeNullable",
+                               {{ts_lv,
+                                 cgen_state_->llInt(scale),
+                                 cgen_state_->llInt(target_sec_scale),
+                                 cgen_state_->inlineIntNull(ctx.int64())}});
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: agree, passing source and target null values

File: omniscidb/QueryEngine/DateTruncate.cpp [hunk 8]
```diff
@@ -330,6 +330,78 @@ DateTruncateHighPrecisionToDateNullable(const int64_t timeval,
...
+      timeval, conersion_scale, target_sec_scale);
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE DEVICE int64_t
+TimestampTruncateHighPrecisionToBiggerTimeNullable(const int64_t timeval,
+                                                   const int64_t conersion_scale,
+                                                   const int64_t target_sec_scale,
+                                                   const int64_t null_val) {
+  if (timeval == null_val) {
+    return null_val;
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/DateTruncate.cpp]: agree, done

File: omniscidb/QueryEngine/CastIR.cpp [hunk 9]
```diff
@@ -159,11 +159,55 @@ llvm::Value* CodeGenerator::codegenCastTimestampToDate(llvm::Value* ts_lv,
...
+  const auto operand_unit = operand_type->as<hdk::ir::TimestampType>()->unit();
+  const auto target_unit = target_type->as<hdk::ir::TimeType>()->unit();
+  const auto target_sec_scale = hdk::ir::unitsPerSecond(target_unit);
+
+  CHECK(ts_lv->getType()->isIntegerTy(64));
+  if (operand_unit < target_unit) {
+    const auto scale =
+        hdk::ir::unitsPerSecond(target_unit) / hdk::ir::unitsPerSecond(operand_unit);
+
+    // codegenCastBetweenIntTypesOverflowChecks(ts_lv, operand_type, target_type, scale);
```
Comment (ienkovich) [inline omniscidb/QueryEngine/CastIR.cpp:175]: Remove comment?

File: omniscidb/QueryEngine/CastIR.cpp [hunk 10]
```diff
@@ -159,11 +159,55 @@ llvm::Value* CodeGenerator::codegenCastTimestampToDate(llvm::Value* ts_lv,
...
+                                 {{ts_lv,
+                                   cgen_state_->llInt(scale),
+                                   cgen_state_->llInt(target_sec_scale),
+                                   cgen_state_->inlineIntNull(operand_type),
+                                   cgen_state_->inlineIntNull(target_type)}});
+  }
+  const auto scale =
+      hdk::ir::unitsPerSecond(operand_unit) / hdk::ir::unitsPerSecond(target_unit);
+
+  // codegenCastBetweenIntTypesOverflowChecks(ts_lv, operand_type, target_type, scale);
```
Comment (ienkovich) [inline omniscidb/QueryEngine/CastIR.cpp:193]: Remove comment?

File: omniscidb/ResultSet/ResultSet.cpp [hunk 11]
```diff
@@ -256,10 +256,13 @@ std::string ResultSet::getStrScalarVal(const ScalarTargetValue& current_scalar,
         if (col_type->isDate()) {
           oss << getStrDateFromSeconds(boost::get<int64_t>(current_scalar));
         } else if (col_type->isTime()) {
-          oss << getStrTimeFromSeconds(boost::get<int64_t>(current_scalar));
+          oss << getStrTStamp(boost::get<int64_t>(current_scalar),
+                              col_type->as<hdk::ir::TimeType>()->unit())
+              << " scalar: " << current_scalar;
```
Comment (ienkovich) [inline omniscidb/ResultSet/ResultSet.cpp:261]: I think this function is used to print result sets in PyHDK, scalar values would be really unexpected there.

File: omniscidb/Tests/QueryBuilderTest.cpp [hunk 12]
```diff
@@ -519,6 +562,180 @@ TEST_F(QueryBuilderTest, Arithmetics) {
   compare_res_data(res, std::vector<int64_t>({0, NULL_BIGINT}));
 }
 
+TEST_F(QueryBuilderTest, TmstmpToTime) {
```
Comment (ienkovich) [inline omniscidb/Tests/QueryBuilderTest.cpp:565]: Please use unabbreviated `TimestampToTime`

File: omniscidb/ResultSet/ResultSet.cpp (hunk 11 above)
Comment (Devjiu) [reply omniscidb/ResultSet/ResultSet.cpp:261]: `getStrTStamp` print all time types as timestamp, it looks like a debug function. If it's for end user we should rewrite it.

File: omniscidb/QueryEngine/CastIR.cpp (hunk 9 above)
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: done

File: omniscidb/QueryEngine/CastIR.cpp (hunk 10 above)
Comment (Devjiu) [reply omniscidb/QueryEngine/CastIR.cpp]: done

File: omniscidb/Tests/QueryBuilderTest.cpp (hunk 12 above)
Comment (Devjiu) [reply omniscidb/Tests/QueryBuilderTest.cpp]: done

File: omniscidb/QueryEngine/DateTruncate.cpp [hunk 13]
```diff
@@ -330,6 +330,64 @@ DateTruncateHighPrecisionToDateNullable(const int64_t timeval,
   return truncate_high_precision_timestamp_to_date(timeval, scale);
 }
 
+namespace {
+// scale is 10^{3,6,9}
+inline DEVICE int64_t
+timestamp_to_time_truncation_upscale(const int64_t timeval,
+                                     const int64_t conersion_scale,
```
[the remark this answers is not in the document]
Comment (Devjiu) [reply omniscidb/QueryEngine/DateTruncate.cpp]: done

**Review by ienkovich — APPROVED** [review]

**Pull request merged**